### PR TITLE
Support MySQL Part 1

### PR DESCRIPTION
**Pull request by HaitaoDeng**

### CODE REVIEW

File: lib/l10n/intl_en.arb [hunk 1]
```diff
@@ -57,5 +57,8 @@
   "topicName": "Topic Name",
   "topics": "Topics",
   "unit": "unit",
-  "versionName": "Version"
+  "versionName": "Version",
+  "database": "Database",
+  "tables": "Tables",
+  "userName": "UserName"
 }
```
Comment (goflutterjava) [inline lib/l10n/intl_en.arb:64]: arb file is sort by lexicographical order

File: lib/l10n/intl_zh.arb [hunk 2]
```diff
@@ -57,5 +57,7 @@
   "topicName": "Topic 名称",
   "topics": "Topic 列表",
   "unit": "单位",
-  "versionName": "版本"
+  "database": "数据库",
```
Comment (goflutterjava) [inline lib/l10n/intl_zh.arb:60]: arb file is sort by lexicographical order

**Comment by HaitaoDeng** [comment]
image
image
image
three Windows

File: lib/module/mysql/const.dart [hunk 3]
```diff
@@ -21,6 +21,6 @@ class MysqlConst {
   static const String infoDb = "information_schema";
   static const String defaultHost = "localhost";
   static const int defaultPort = 3306;
-  static const String defaultUsername = "hzj";
-  static const String defaultPassword = "Mysql@123";
+  static const String defaultUsername = "root";
+  static const String defaultPassword = "123456";
```
Comment (goflutterjava) [inline lib/module/mysql/const.dart:25]: no need to commit

**Review by goflutterjava — CHANGES_REQUESTED** [review]

File: lib/ui/mysql/mysql_page.dart [hunk 4]
```diff
@@ -77,7 +78,7 @@ class _MysqlPageState extends State<MysqlPage> {
             rows: vm.instances
                 .map((itemRow) => DataRow(
                         onSelectChanged: (bool? selected) {
-                          Navigator.pushNamed(context, PageRouteConst.MysqlInstance, arguments: itemRow.deepCopy());
+                          Navigator.of(context).pushNamed(PageRouteConst.MysqlInstance, arguments: itemRow.deepCopy());
```
Comment (goflutterjava) [inline lib/ui/mysql/mysql_page.dart:81]: why change this?

File: lib/ui/mysql/mysql_page.dart [hunk 5]
```diff
@@ -25,6 +25,7 @@ import 'package:paas_dashboard_flutter/ui/util/form_util.dart';
 import 'package:paas_dashboard_flutter/vm/mysql/mysql_instance_list_view_model.dart';
 import 'package:provider/provider.dart';
 
+/// MySQL实例列表页面
```
Comment (goflutterjava) [inline lib/ui/mysql/mysql_page.dart:28]: use english annotation

File: lib/ui/mysql/widget/mysql_database.dart [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import 'package:flutter/material.dart';
+import 'package:paas_dashboard_flutter/generated/l10n.dart';
+import 'package:paas_dashboard_flutter/route/page_route_const.dart';
+import 'package:paas_dashboard_flutter/ui/util/exception_util.dart';
+import 'package:paas_dashboard_flutter/ui/util/spinner_util.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_database_view_model.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_table_view_model.dart';
+import 'package:provider/provider.dart';
+
+/// mysql 数据库列表
```
Comment (goflutterjava) [inline lib/ui/mysql/widget/mysql_database.dart:29]: use english annotation

File: lib/ui/mysql/widget/mysql_tables.dart [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            title: Text('Mysql ${vm.name} -> ${vm.getDbname()} DB'),
+            bottom: TabBar(
+              tabs: [
+                Tab(text: "Tables"),
+              ],
+            ),
+          ),
+          body: body,
+        ));
+    // return body;
```
Comment (goflutterjava) [inline lib/ui/mysql/widget/mysql_tables.dart:127]: nits.

File: lib/vm/mysql/mysql_database_view_model.dart [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import 'package:paas_dashboard_flutter/api/mysql/mysql_databases_api.dart';
+import 'package:paas_dashboard_flutter/module/mysql/mysql_database.dart';
+import 'package:paas_dashboard_flutter/persistent/po/mysql_instance_po.dart';
+
+import '../base_load_list_page_view_model.dart';
```
Comment (goflutterjava) [inline lib/vm/mysql/mysql_database_view_model.dart:24]: use absolute import path

File: lib/vm/mysql/mysql_user_view_model.dart [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import '../base_load_list_page_view_model.dart';
```
Comment (goflutterjava) [inline lib/vm/mysql/mysql_user_view_model.dart:20]: use absolute import path.

File: lib/vm/mysql/mysql_table_data_view_model.dart [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import 'package:flutter/material.dart';
+import 'package:paas_dashboard_flutter/api/mysql/mysql_databases_api.dart';
+import 'package:paas_dashboard_flutter/module/mysql/mysql_sql_result.dart';
+import 'package:paas_dashboard_flutter/persistent/po/mysql_instance_po.dart';
+
+import '../base_load_list_page_view_model.dart';
```
Comment (goflutterjava) [inline lib/vm/mysql/mysql_table_data_view_model.dart:25]: use absolute import

File: lib/vm/mysql/mysql_table_data_view_model.dart [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+class MysqlTableDetailViewModel extends BaseLoadListPageViewModel<Object> {
```
Comment (goflutterjava) [inline lib/vm/mysql/mysql_table_data_view_model.dart:27]: why we use `Object` here?

File: lib/l10n/intl_zh.arb [hunk 12]
```diff
@@ -57,5 +57,7 @@
...
+  "tables": "数据库表表",
```
Comment (zxJin-x) [inline lib/l10n/intl_zh.arb:61]: 数据库表表？

File: lib/ui/mysql/widget/mysql_tables.dart [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+import 'package:paas_dashboard_flutter/generated/l10n.dart';
+import 'package:paas_dashboard_flutter/route/page_route_const.dart';
+import 'package:paas_dashboard_flutter/ui/component/searchable_title.dart';
+import 'package:paas_dashboard_flutter/ui/util/exception_util.dart';
+import 'package:paas_dashboard_flutter/ui/util/spinner_util.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_table_data_view_model.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_table_view_model.dart';
+import 'package:provider/provider.dart';
+
+/// mysql 数据库表list界面
```
Comment (goflutterjava) [inline lib/ui/mysql/widget/mysql_tables.dart:30]: use english annotation

File: lib/ui/mysql/widget/mysql_table_detail.dart [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import 'package:flutter/material.dart';
+import 'package:paas_dashboard_flutter/generated/l10n.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_table_data_view_model.dart';
+import 'package:provider/provider.dart';
+
+/// MySQL 表数据展示
```
Comment (goflutterjava) [inline lib/ui/mysql/widget/mysql_table_detail.dart:25]: please use english annotation

**Review by zxJin-x — CHANGES_REQUESTED** [review]

File: lib/ui/mysql/widget/mysql_database.dart [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        Container(
+          height: 50,
+          child: ListView(
+            scrollDirection: Axis.horizontal,
+            shrinkWrap: true,
+            children: [refreshButton],
+          ),
+        ),
+        Text(
+          'Mysql Database',
```
Comment (zxJin-x) [inline lib/ui/mysql/widget/mysql_database.dart:98]: image
是不是有点重复了

File: lib/ui/mysql/widget/mysql_tables.dart [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        Container(
+          height: 50,
+          child: ListView(
+            scrollDirection: Axis.horizontal,
+            shrinkWrap: true,
+            children: [refreshButton],
+          ),
+        ),
+        Text(
+          'Mysql tables',
```
Comment (zxJin-x) [inline lib/ui/mysql/widget/mysql_tables.dart:106]: 做国际化

**Review by zxJin-x — CHANGES_REQUESTED** [review]

File: lib/ui/mysql/widget/mysql_tables.dart [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+          style: TextStyle(fontSize: 22),
+        ),
+        SearchableTitle("database list", "search by database name", searchTextController),
```
Comment (zxJin-x) [inline lib/ui/mysql/widget/mysql_tables.dart:109]: database list做国际化

File: lib/ui/mysql/widget/mysql_user.dart [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+        Container(
+          height: 50,
+          child: ListView(
+            scrollDirection: Axis.horizontal,
+            shrinkWrap: true,
+            children: [refreshButton],
+          ),
+        ),
+        Text(
+          'Mysql User Name',
```
Comment (zxJin-x) [inline lib/ui/mysql/widget/mysql_user.dart:87]: 国际化

File: lib/api/mysql/mysql_databases_api.dart [hunk 19]
```diff
@@ -17,10 +17,68 @@
...
+import 'package:sprintf/sprintf.dart';
 
 class MysqlDatabaseApi {
+  static const String SELECT_ALL = "select * from %s limit 1000";
+
   static Future<List<DatabaseResp>> getDatabaseList(String host, int port, String username, String password) async {
-    throw UnimplementedError();
+    final setting = new ConnectionSettings(host: host, port: port, user: username, password: password);
+    final MySqlConnection conn = await MySqlConnection.connect(setting);
+    var rr = await conn.query('show databases');
```
Comment (jiaob) [inline lib/api/mysql/mysql_databases_api.dart:33]: The variable name is incorrect.

**Comment by goflutterjava** [comment]
@HaitaoDeng Please ping me if you have addressed all problems
cc @jiaob @zxJin-x @fu-turer

File: lib/main.dart [hunk 20]
```diff
@@ -170,6 +173,19 @@ class MyApp extends StatelessWidget {
           final args = settings.arguments as SqlViewModel;
           return RouteGen.sqlExecute(args);
         }
+        if (settings.name == PageRouteConst.MysqlInstance) {
```
Comment (goflutterjava) [inline lib/main.dart:176]: suggest put mysql under mongo. lexicographical order

File: lib/route/route_gen.dart [hunk 21]
```diff
@@ -135,4 +141,28 @@ class RouteGen {
               child: SqlExecuteScreen(),
             ));
   }
+
+  static Route mysqlInstance(MysqlInstanceViewModel viewModel) {
```
Comment (goflutterjava) [inline lib/route/route_gen.dart:145]: same, under mongo, lexicographical

File: lib/ui/mysql/mysql_instance.dart [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+//
+
+import 'package:flutter/material.dart';
+import 'package:paas_dashboard_flutter/ui/mysql/widget/mysql_database.dart';
+import 'package:paas_dashboard_flutter/ui/mysql/widget/mysql_user.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_database_view_model.dart';
+import 'package:paas_dashboard_flutter/vm/mysql/mysql_instance_view_model.dart';
+import 'package:provider/provider.dart';
+
+/// MySQL单个实例，数据库列表页面
```
Comment (goflutterjava) [inline lib/ui/mysql/mysql_instance.dart]: english annotation

File: lib/ui/mysql/mysql_instance.dart [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+    return DefaultTabController(
+      length: 2,
+      child: Scaffold(
+        appBar: AppBar(
+          title: Text('Mysql ${vm.name} db Dashboard'),
+          bottom: TabBar(tabs: [
+            Tab(
+              text: "databases",
+            ),
+            Tab(text: "Details"),
```
Comment (goflutterjava) [inline lib/ui/mysql/mysql_instance.dart:51]: i10n

File: lib/ui/mysql/mysql_instance.dart [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    MysqlInstanceViewModel vm = Provider.of<MysqlInstanceViewModel>(context);
+
+    return DefaultTabController(
+      length: 2,
+      child: Scaffold(
+        appBar: AppBar(
+          title: Text('Mysql ${vm.name} db Dashboard'),
+          bottom: TabBar(tabs: [
+            Tab(
+              text: "databases",
```
Comment (goflutterjava) [inline lib/ui/mysql/mysql_instance.dart:49]: i10n

File: lib/vm/mysql/mysql_instance_view_model.dart [hunk 25]
```diff
@@ -17,9 +17,12 @@
 // under the License.
 //
 
+import 'package:paas_dashboard_flutter/api/mysql/mysql_databases_api.dart';
 import 'package:paas_dashboard_flutter/persistent/po/mysql_instance_po.dart';
 
-class MysqlInstanceViewModel {
+import '../base_load_list_page_view_model.dart';
```
Comment (goflutterjava) [inline lib/vm/mysql/mysql_instance_view_model.dart:23]: absoulte import

File: lib/vm/mysql/mysql_table_view_model.dart [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import 'package:paas_dashboard_flutter/api/mysql/mysql_databases_api.dart';
+import 'package:paas_dashboard_flutter/module/mysql/mysql_table.dart';
+import 'package:paas_dashboard_flutter/persistent/po/mysql_instance_po.dart';
+
+import '../base_load_list_page_view_model.dart';
```
Comment (goflutterjava) [inline lib/vm/mysql/mysql_table_view_model.dart:24]: absoulte import

**Review by goflutterjava — CHANGES_REQUESTED** [review]

File: lib/main.dart [hunk 27]
```diff
@@ -170,6 +185,7 @@ class MyApp extends StatelessWidget {
           final args = settings.arguments as SqlViewModel;
           return RouteGen.sqlExecute(args);
         }
+
```
Comment (goflutterjava) [inline lib/main.dart:188]: unneeded space line

**Comment by goflutterjava** [comment]
@HaitaoDeng are this changes need for format ?
image

**Comment by HaitaoDeng** [comment]
> @HaitaoDeng are this changes need for format ? image

misoperation will not affect?

**Review by goflutterjava — APPROVED** [review]

**Pull request merged**